### ISSUE TITLE
Ryu Shim and zeromq deadlock

### ISSUE THREAD

**Issue by doriguzzi** [fp7-netide/Engine#24]
I'm adding the zeromq support to the Ryu shim and backend. I've pushed some initial code into the ryu-protocol-v1.1 branch.
The strange thing is that when I start the thread that handles the incoming messages, the main thread (the Ryu framework) stops working. The problem comes from line 156 of ryu-shim.py

```
message = self.socket.recv_multipart()
```

If I comment that line, the Ryu shim can receive messages from Mininet and send them to the dummy core (as expected). However, it cannot receive anything (of course) from the core since the recv_multipart is commented. It does not seem a threading locking problem, since if I replace that line with a long sleep, the main thread works perfectly.

It seems like `self.socket.recv_multipart()` is generating a conflict with the sockets that handle the connection with the OF devices, but I'm not sure at all.

Any suggestion?

**Comment by ElisaRojas** [comment]
Hi Roberto. Do you know if they are trying to use the same port reserved for OF? That's the only thing I can think of, but I guess you've already checked that. Maybe we can exclude those port from ZeroMQ somehow.

Can you share the error line you get (if any)?

**Comment by doriguzzi** [comment]
Hi Elisa. In this case the shim is a client, therefore it does not try to bind any port. What it does is: (i)connecting to the ProxyCore (localhost,41414), (ii) setting up a separate thread in order to handle the reception of the messages from the Core without blocking the rest of the Shim.

Unfortunately there are no errors. What happens is that everything works fine until the `self.socket.recv_multipart()` is called. If I put a sleep before, I can see the messages coming from mininet and that are correctly forwarded to the Core. When the sleep expires, the recv start waiting for the messages from the Core, but the rest of Ryu stops working. It seems to me that the partt of Ryu which is affected is the Application Manager, but I need to inspect the code a little bit deeper.
Another effect: if I stop the blocked Ryu Shim by pressing CTRL+C, I actually stop the recv thread and the Ryu Shim start working again. So, I need to press CTRL+C again to stop the Shim.

**Comment by TimVinkemeier** [comment]
Hi Roberto,

I do not yet know the exact problem, but I found something that you have to change anyway. In line 266:
(`self.CoreConnection.socket.send(msg_to_send)`) you access the socket from a thread that is not the owner of the socket. ZeroMQ sockets are NOT threadsafe, so this access is not supported. I don't know if that changes anything, but you definitely should resolve that. Have a look at the Java example client and adapt that pattern (it is the recommended pattern from the ZeroMQ guide).

Best,
Tim

**Comment by TimVinkemeier** [comment]
Same goes for gracefully shutting down the receive-thread. It is not recommended to directly call `recv_multipart`, because it is a blocking operation on the socket (as far as I know). Use a Poller, as shown in the Java client example. I should likely change the CoreClient.py file in the SVN, since it does not use the recommended pattern (but works, interestingly :/)

**Comment by doriguzzi** [comment]
What I did was basically using the CoreClient.py that I found in the SVN. That sample program uses the `send` method outside the thread and it works perfectly with the ProxyCore.
However, I'm pretty sure that it is not the cause of the problem.

Regarding the usage of the `recv_multipart` method, this is the typical way of waiting messages. It blocks until a message arrives (like with "legacy" sockets). In this way you don't have to keep on polling the channel. However I will take a look to the Poller method.

**Comment by doriguzzi** [comment]
I close this issue because I solved it. The problem was due to the eventlet library (http://eventlet.net/) used by Ryu to handle the applications. They say that the library is compatible with Python native threads, but I found that the listening thread had the same process id of the main process. So, everything was running inside the same thread and, therefore, the recv_multipart blocked the whole Ryu controller.
In fact this library is based on "green threads" (http://eventlet.net/doc/basic_usage.html) and, based on this concept,  they also provide a "green" version of the zmq interface that solves the problem.
